### PR TITLE
Debugger Refactor #2: `DebuggerArgs`

### PR DESCRIPTION
**Pull request by iFrostizz**
Second and last part of https://github.com/foundry-rs/foundry/pull/5547 (sorry, cannot do smoller)
Adding the new `DebuggerArgs` struct that all part of the code should comply with,
Removing the `run_debugger` function that is ran from `forge script`
Simpler way of mapping contract sources, see `ui::debugger::ContractSources` type

### CODE REVIEW

**Review by Evalir — CHANGES_REQUESTED** [review]
nice! let's fix conflicts before doing a deeper review

**Comment by iFrostizz** [comment]
yep doing that now sorry sir

**Comment by iFrostizz** [comment]
did a self review, looks good to me @Evalir 
there are less changes than it seems in the diff, git is not smart enough to show the correct diffs for `lib.rs` and make it look like there are many.

File: crates/common/src/compile.rs
Comment (Evalir) [inline crates/common/src/compile.rs:404]: Hmm, let's either make an alias type for `u32` or comment what the returned tuple means—I had to read the fn source to understand that it was a file_id

File: crates/common/src/contracts.rs
Comment (Evalir) [inline crates/common/src/contracts.rs:278]: Let's not unwrap here—rather return a Result

File: crates/evm/src/fuzz/mod.rs
Comment (Evalir) [inline crates/evm/src/fuzz/mod.rs:248]: let's add docs

File: crates/evm/src/trace/identifier/etherscan.rs
Comment (Evalir) [inline crates/evm/src/trace/identifier/etherscan.rs:64]: should we resolve this todo?

File: crates/evm/src/trace/identifier/etherscan.rs
Comment (Evalir) [inline crates/evm/src/trace/identifier/etherscan.rs:70]: This shuffles ordering—is ordering an issue here?

File: crates/forge/bin/cmd/script/build.rs
Comment (Evalir) [inline crates/forge/bin/cmd/script/build.rs:47]: let's avoid unwrapping

File: crates/forge/bin/cmd/script/build.rs
Comment (Evalir) [inline crates/forge/bin/cmd/script/build.rs:283]: Almost sure this might cause linking problems, ordering might be important here

File: crates/forge/bin/cmd/script/cmd.rs
Comment (Evalir) [inline crates/forge/bin/cmd/script/cmd.rs:89]: `unwrap_or_else` might be more performant here

File: crates/forge/src/runner.rs
Comment (Evalir) [inline crates/forge/src/runner.rs:543]: todo here

**Review by Evalir — CHANGES_REQUESTED** [review]
need to look at this more closely, but some preliminary comments

File: crates/common/src/compile.rs
Comment (iFrostizz) [reply crates/common/src/compile.rs:404]: Yep, good idea thanks

File: crates/evm/src/trace/identifier/etherscan.rs
Comment (iFrostizz) [reply crates/evm/src/trace/identifier/etherscan.rs:70]: We don't need ordering, so there should be better performances with a `HashMap`
Can leave as-is if not appriopriate for this PR

File: crates/forge/bin/cmd/script/build.rs
Comment (iFrostizz) [reply crates/forge/bin/cmd/script/build.rs:47]: right, thanks!

File: crates/forge/bin/cmd/script/build.rs
Comment (iFrostizz) [reply crates/forge/bin/cmd/script/build.rs:283]: Could you please elaborate ? I'm not sure why ordering would matter because we map the sources from the artifact Id https://github.com/foundry-rs/foundry/pull/5753/files/466d24a4be210ff3be7b7eedb8e62c76e4630d6e#diff-fdebd782f78bd93139131fe4f614d8ab47564bceef84dd9a009483b0c5ac7b41R42

File: crates/forge/src/runner.rs
Comment (iFrostizz) [reply crates/forge/src/runner.rs:543]: I don't think that debugging traces were collected before for invariant testing, could we offload this for another PR ?

File: crates/common/src/contracts.rs
Comment (iFrostizz) [reply crates/common/src/contracts.rs:278]: yep, def cleaner, thanks!

File: crates/evm/src/fuzz/mod.rs
Comment (iFrostizz) [reply crates/evm/src/fuzz/mod.rs:248]: Right, thank you

File: crates/evm/src/trace/identifier/etherscan.rs
Comment (iFrostizz) [reply crates/evm/src/trace/identifier/etherscan.rs:64]: Yes, moved it to `crates/common/src/compile.rs` ! Thanks

File: crates/forge/bin/cmd/script/cmd.rs
Comment (iFrostizz) [reply crates/forge/bin/cmd/script/cmd.rs:89]: `unwrap_or_default` even better, wdyt ?

File: crates/forge/src/runner.rs
Comment (iFrostizz) [reply crates/forge/src/runner.rs]: Added a better desc on the TODO!

File: crates/common/src/compile.rs
Comment (mattsse) [inline crates/common/src/compile.rs:402]: please no public type aliases for primtive values. either make it private or add a new wrapper type.

and move to top of file

File: crates/common/src/compile.rs
Comment (mattsse) [inline crates/common/src/compile.rs:406]: please try to avoid these aliases, I really dislike them.

this should just be a new helper type.

File: crates/forge/bin/cmd/script/build.rs
Comment (mattsse) [inline crates/forge/bin/cmd/script/build.rs:36]: this is one reason why I dislike this so much

**Review by mattsse — CHANGES_REQUESTED** [review]
Since we're refactoring this to make it better, I'm being extra nitpicky here

File: crates/common/src/compile.rs
Comment (iFrostizz) [reply crates/common/src/compile.rs:406]: Can you please be extra explicit on what is a wrapper type ?
I found this alias convenient to have one place where it's defined and the reader can just goto definition and read the docs, instead of copy pasting them everywhere this is used and have this two-dimensional map.

Comment (mattsse) [reply crates/common/src/compile.rs:406]: > Can you please be extra explicit on what is a new helper type ?

a new type that encapsulates this map.

this has a few advantages even if it requires more code,
1 type only
better debugging
custom helper functions

File: crates/common/src/compile.rs
Comment (iFrostizz) [reply crates/common/src/compile.rs:402]: I think that it's reasonable to just drop this type

File: crates/common/src/compile.rs
Comment (iFrostizz) [reply crates/common/src/compile.rs:406]: I see, thanks

File: crates/forge/bin/cmd/script/build.rs
Comment (iFrostizz) [reply crates/forge/bin/cmd/script/build.rs]: Understood

**Comment by iFrostizz** [comment]
> Since we're refactoring this to make it better, I'm being extra nitpicky here

This is good

File: crates/forge/bin/cmd/script/cmd.rs
Comment (Evalir) [reply crates/forge/bin/cmd/script/cmd.rs]: yep sounds good

File: crates/forge/src/runner.rs
Comment (Evalir) [reply crates/forge/src/runner.rs]: ah cool! makes sense then.

File: crates/common/src/compile.rs
Comment (Evalir) [inline crates/common/src/compile.rs:174]:
```suggestion
/// Map over artifacts contract sources name -> file_id -> (source, contract)
```

File: crates/forge/bin/cmd/test/mod.rs
Comment (Evalir) [inline crates/forge/bin/cmd/test/mod.rs:329]: let's write docs for these

File: crates/forge/src/multi_runner.rs
Comment (Evalir) [inline crates/forge/src/multi_runner.rs:76]: let's also add docs here

File: crates/forge/src/runner.rs
Comment (Evalir) [inline crates/forge/src/runner.rs:591]: will we resolve this?

**Review by Evalir — CHANGES_REQUESTED** [review]
a few more comments, need to do another pass on the linking section—sorry we're being nitpicky, we just really wanna improve this! great work so far

File: crates/forge/src/runner.rs
Comment (iFrostizz) [reply crates/forge/src/runner.rs:591]: I don't know, I'm not sure how to handle this because everything has been shaped for one `calldata` and here it messes things up to have a sequence of calls

Comment (iFrostizz) [reply crates/forge/src/runner.rs:591]: Could we do that in #3 ? Probably modify the `single_fuzz` to take in a `Vec<calldata, value>` to support sequence calls. Just noticed that the value is defaulted to 0 right now as well, might make it a bug

Comment (iFrostizz) [reply crates/forge/src/runner.rs:591]: I'm just not familiar how we can force the fuzzer to construct a sequence of calls

File: crates/common/src/compile.rs
Comment (mattsse) [inline crates/common/src/compile.rs:408]: this needs docs that this returns the fileid of the solidity artifact.

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, defer to @Evalir to get this over the line

File: crates/common/src/compile.rs
Comment (iFrostizz) [reply crates/common/src/compile.rs:409]: Done, thanks!

Comment (iFrostizz) [reply crates/common/src/compile.rs:409]: Done thanks!

**Comment by iFrostizz** [comment]
Friendly ping @Evalir

**Review by mattsse — CHANGES_REQUESTED** [review]
lgtm pending conflicts, sorry about that

**Review by Evalir — CHANGES_REQUESTED** [review]
yep agreed with mattsse_—sorry for the rebasing work needed here

**Comment by iFrostizz** [comment]
No it's all G, thanks for the thorough review !

**Review by Evalir — APPROVED** [review]
gg, sweet, happy that this is over the line! (let's merge on monday just in case)

**Pull request merged**
